Sources/Algorithms/Stride.swift

```diff
@@ -0,0 +1,224 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+//===----------------------------------------------------------------------===//
+// striding(by:)
+//===----------------------------------------------------------------------===//
+
+extension Sequence {
+  /// Returns a sequence stepping through the elements every `step` starting
+  /// at the first value. Any remainders of the stride will be trimmed.
+  ///
+  ///     (0...10).striding(by: 2) // == [0, 2, 4, 6, 8, 10]
+  ///     (0...10).striding(by: 3) // == [0, 3, 6, 9]
+  ///
+  /// - Complexity: O(1). Access to successive values is O(1) if the
+  /// collection conforms to `RandomAccessCollection`; otherwise,
+  /// O(_k_), where _k_ is the striding `step`.
+  ///
+  /// - Parameter step: The amount to step with each iteration.
+  /// - Returns: Returns a sequence or collection for stepping through the
+  /// elements by the specified amount.
+  public func striding(by step: Int) -> Stride<Self> {
+    Stride(base: self, stride: step)
+  }
+}
+
+public struct Stride<Base: Sequence> {
+  
+  let base: Base
+  let stride: Int
+  
+  init(base: Base, stride: Int) {
+    precondition(stride > 0, "striding must be greater than zero")
+    self.base = base
+    self.stride = stride
+  }
+}
+
+extension Stride {
+  public func striding(by step: Int) -> Self {
+    Stride(base: base, stride: stride * step)
+  }
+}
+
+extension Stride: Sequence {
+  
+  public struct Iterator: IteratorProtocol {
+    
+    var iterator: Base.Iterator
+    let stride: Int
+    
+    public mutating func next() -> Base.Element? {
+      defer {
+        for _ in 0..<(stride - 1) {
+          guard iterator.next() != nil else { break }
+        }
+      }
+      return iterator.next()
```

Comment:
This always immediately computes the `stride - 1` elements after the one being returned. Would it be better to compute them only when the element after them is being returned, instead?

Reply:
Yes it does. I had another implementation which deferred the compute, it does add more storage onto the iterator. With the worst case being an extra O(k-1), if `k` is large and unbounded it may cause problems so I think I will look again toward this approach

```swift
extension Stride: Sequence {
  
  public struct Iterator: IteratorProtocol {

    var iterator: Base.Iterator
    let stride: Int
    var striding: Bool = false
    
    public mutating func next() -> Base.Element? {
      
      guard striding else {
        striding = true
        return iterator.next()
      }
      
      for _ in 0..<(stride - 1) {
        guard iterator.next() != nil else { break }
      }
      return iterator.next()
    }
  }
  
  public func makeIterator() -> Stride<Base>.Iterator {
    Iterator(iterator: base.makeIterator(), stride: stride)
  }
}
```

Reply:
This looks reasonable, iterators are almost always temporary so it should be nog big deal to add a `Bool` to it.